node/lib/schema/OcfSchemaReader.ts

```diff
@@ -1,40 +1,380 @@
+     * @returns {Promise<ThingSchema>} Schema parsed from the RAML and JSON files
+     */
     public static async readThingSchemaFromFilesAsync(
             ramlFilePath: string): Promise<ThingSchema> {
-        let raml = await fs.readFile(ramlFilePath, "utf8");
+        let ramlSchema: (raml.api08.Api | raml.api10.Api) = await raml.loadApi(ramlFilePath);
+        let thingSchema: ThingSchema = OcfSchemaReader.ramlToThingSchema(ramlSchema);
 
-        // TODO: Scan for `!include *.json` lines in the RAML, then load JSON files (async).
-        // TODO: Scan for `$ref` objects in each JSON, then recursively load references (async).
+        let schemaResolver: (fileName: string) => Promise<string> =
+            OcfSchemaReader.resolveJsonSchemaAsync.bind(null, path.dirname(ramlFilePath));
 
-        throw new Error("not implemented");
+        for (let methodIndex = 0; methodIndex < thingSchema.methods.length; methodIndex++) {
+            let method = thingSchema.methods[methodIndex];
+            for (let paramIndex = 0; paramIndex < method.parameters.length; paramIndex++) {
+                let parameter = method.parameters[paramIndex];
+                method.parameters[paramIndex] = {
+                    isOut: parameter.isOut,
+                    name: parameter.name,
+                    parameterType: await JsonSchema.resolveReferencesAsync(
+                        parameter.parameterType, schemaResolver),
+                };
+            }
+        }
+
+        return thingSchema;
     }
 
+    /**
+     * Reads a thing schema from an OCF RAML string and supporting JSON schema strings.
+     *
+     * @param {string} raml  RAML contents
+     * @param {{[fileName: string]: string}} json  Map from JSON file names to JSON contents
+     * @returns {ThingSchema} Schema parsed from the RAML and JSON strings
+     */
     public static readThingSchema(
             raml: string, json: {[fileName: string]: string}): ThingSchema {
+        // TODO: Call raml.parseRAMLSync() and pass in a resolver for JSON includes.
         throw new Error("not implemented");
     }
+
+    /**
+     * Convert a parsed RAML API to a thing schema.
+     *
+     * Note this code (mostly) supports RAML v0.8 and v1.0 APIs.
+     * (OCF currently uses 0.8, so RAML 1.0 support is untested at this point.)
+     */
+    private static ramlToThingSchema(
+            ramlSchema: (raml.api08.Api | raml.api10.Api)): ThingSchema {
+        let schemaName = ramlSchema.title();
+        let description: string | undefined;
+        let methods: ThingMethod[] = [];
+
+        let ramlResources: Array<raml.api08.Resource | raml.api10.Resource> =
+                ramlSchema.resources();
+        ramlResources.forEach(ramlResource => {
+            // Remove a leading slash from the resource relative URI.
+            // Replace any other slashes, dots, or dashes with underscores.
+            let resourceName: string = ramlResource.relativeUri().value()
```

Comment:
[AllJoyn member names aren't allowed to have any of those characters.](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names)

> Member names must only contain the ASCII characters "[A-Z][a-z][0-9]_" and may not begin with a digit.

---
In reply to: [75385990](https://github.com/openT2T/opent2t/pull/23#discussion_r75385990) [](ancestors = 75385990)